### ISSUE TITLE
Investigate better ways to migrate between network sizes

### ISSUE THREAD

**Issue by Mardak** [gcp/leela-zero#648]
Splitting out a discussion for @gcp's comment in https://github.com/gcp/leela-zero/issues/591#issuecomment-357496732
> Bootstrapping 6x128 was (very!) unsuccessful so far. From eyeballing the results (and experience with supervised learning), the value network on those is overfit. It's possible to control this, but training new networks is going to take a week or so in any case. Just forcing the clients to play games with any of those (which are pretty much known to be in a bad state) seems a waste of resources. I'd rather we do testing and see if we can produce a better network from the data we have. We can still start a new run from that instead of 0.

It seems likely that multiple network size migrations will happen in the future -- potentially to both larger and smaller (e.g., trying to compress for faster play on lower power devices).

I don't have a good grasp of the research done so far, so I'm not sure if this is new territory worth doing research or if it's something existing that just needs to be implemented. If it is the latter case, it could still be an interesting research area of ways to do it in a distributed fashion.

Do people have experience or ideas?

**Comment by odeint** [comment]
I know people prune and/or quantise network weights to improve performance, but that's usually within the framework of the same network architecture. Going to a new architecture can happen by re-training on data generated by the old one, but I'm not aware of a more direct method. There is some talk of adding identity layers to make a known network deeper - at least in theory it should be possible to make a network deeper without degrading the accuracy. Shallower is probably hard (because you need to selectively remove data), as is changing the filters (it's too different to use the old weights in any direct way).

**Comment by RavnaBergsndot** [comment]
The trained 6x128 networks, without fine hyperparameter tuning, achieved the ELOs that are only 300-400 less than a proven, well-adjusted self-playing-training pipeline. I would call it rather successful.

**Comment by ashinpan** [comment]
Why don't we try the technique from a (2016) Google paper titled "Net2Net: Accelerating Learning via Knowledge Transfer" <https://arxiv.org/abs/1511.05641>? 

If we follow that method, the way forward would be to embed the best available 5x64 network in a new 6x128 network, and train it on *its own" self-play games. 

At the starting point, elo would be rather lower than the best network, but it would still be better than starting from complete random games.

**Comment by ashinpan** [comment]
@RavnaBergsndot "The trained 6x128 networks, without fine hyperparameter tuning, achieved the ELOs that are only 300-400 less than a proven, well-adjusted self-playing-training pipeline. I would call it rather successful." I agree with you when I come to think about it.

**Comment by billyswong** [comment]
If we wish a 6x128 network that is as good as our best 5x64 network or even better, my idea would be to be simulate the self-playing-training process, reusing the self-play games as if they are played by the 6x128 net, but do the training-testing-promotion stepwise by real distributed clients similar to what we did to 5x64 net, including all the lower-learning-rate-when-no-progress mechanism.

**Comment by ashinpan** [comment]
@billyswong Wasn't it how gcp train his 6x128 networks?

**Comment by billyswong** [comment]
I don't think he did the stepwise training-testing-promotion, probably just pouring the game records in a usual supervised training mannr. Even if there is stepwise training-testing-promotion, it must be nowhere close to the intensity of what we are doing to the current 5x64 net.

**Comment by gcp** [comment]
>reusing the self-play games as if they are played by the 6x128 net, but do the training-testing-promotion stepwise by real distributed clients similar to what we did to 5x64 net

The problem is that there's just no guarantee you will ever get a "PASS" on the same training window of data that produced a PASS for the 5x64 network (because 6x128 is bigger and can overfit differently).

If you can't guarantee a PASS on the same window, I don't see how you can have a "testing-promotion" cycle that stays in sync. And if you can't guarantee that cycle, what's the point of the "testing-promotion" part?

**Comment by billyswong** [comment]
My naive thought: Train the first 250k/500k games from 5x64 net to a 6x128 net. Treat it as if it is the "current best". Introduce games bit by bit from 5x64 net and train new 6x128 nets. Test new nets against the trained "current best" 6x128 net. Repeat until promotion occur. Promoted net become the new "current best" and then repeat again. If no promotion occur for the whole window or 2/3 window, lower the learning rate.

Since AGZ used 500k window, I thought we wouldn't need any window bigger than that...

> If you can't guarantee a PASS on the same window, I don't see how you can have a "testing-promotion" cycle that stays in sync.

Do you mean the "lower learning rate" part is not in sync? I didn't know we need them to be perfectly in sync for my naive idea to work. Thanks. I am still hoping it don't need to be in perfect sync, maybe a forced delay of learning rate decrease until the smaller net which played the self-play games decrease it first and still no promotion for another one window or 2/3 window.

**Comment by Marcin1960** [comment]
"I don't understand the logic in the suggestion, nor how it could work in practice."

Freeze/save/forget  the 5x64 progress and regress to the least bad  6x128 net and make it the new "best" net to restart from there. Then give it a week or two to see if it can move up on its own.

If it does, it might eventually reach the level of the best 5x64 and further, If not one can always return to the frozen present state and think about other solutions.

**Comment by gcp** [comment]
>regress to the least bad 6x128 net and make it the new "best" net to restart from there

This is a reasonable method but I do think it should be possible to get a 6x128 to be at least be very close to the 5x64's. The current ones are so much worse that they indicate training issues, IMHO.

**Comment by Marcin1960** [comment]
"This is a reasonable method but I do think it should be possible to get a 6x128 to be at least be very close to the 5x64's. "

That is OK, perhaps it needs time to catch up.

**Comment by Marcin1960** [comment]
Or at least one train a few 6x128 networks and see if one of them can come a little closer. Only a small fraction of new 5x64 can do it.

**Comment by Marcin1960** [comment]
I am a biologist specialized in evolution and ecosystems. As a programmer I am only 8 kyu :), but I have some general understanding and intuition of advancing over-complex structures. They do not go smoothly, Without mass extinctions probably we would not have Mesozoic or Cenozoic.   

Mammals were older than dinosaurs, they just needed a handicap in the right moment.

**Comment by marcocalignano** [comment]
@gcp here there is an implementation for the article @ashinpan referred:
https://github.com/rstudio/keras/blob/master/vignettes/examples/mnist_net2net.py
I am still reading the code to understand if we can use it, but maybe someone can beat me to that ;)

**Comment by Marcin1960** [comment]
Have you seen it ? !

ae4921352.48M+8k VS 0db82470, 7 : 1 (87.50%)

100 points to Gian-Carlo?

**Comment by Dorus** [comment]
ae492135 VS 0db82470 | 16 : 20 (44.44%)
-- | --

Do not celebrate too early :)

**Comment by Mardak** [comment]
Focusing on knowledge transfer as opposed to distilling to smaller networks, a quick read through related works seems to point to methods that speed up training:
https://scholar.google.com/scholar?cites=2686528528183414981

Would it make sense to focus on just one of making the network deeper or wider as opposed to doing both at the same time?

E.g., from @odeint's comment of adding an identity layer to be one deeper and train then evaluate? With so many clients available for running matches, perhaps many +8k networks can be tested at 6x64 as it sounds like the value of additional self play games may be limited anyway? (Using 8k as it's fastest to train and has historically similar win rates as 16k, 32k, and 64k. So just pushing through as many 6x64 network evaluations to see how it behaves.)

**Comment by gcp** [comment]
>E.g., from @odeint's comment of adding an identity layer to be one deeper and train then evaluate?

I think the problem of an identity layer is that it is symmetric, so it can't learn?

(Not very familiar with the research here)

**Comment by gcp** [comment]
>Test new nets against the trained "current best" 6x128 net. Repeat until promotion occur. Promoted net become the new "current best" and then repeat again. If no promotion occur for the whole window or 2/3 window, lower the learning rate.

I objected to this because (as said), there's no guarantee promotion will ever occur. 6x128 is bigger and might need more data to move around.

**Comment by marcocalignano** [comment]
@gcp have a look at this https://github.com/paengs/Net2Net. this is using numpy and tensorflow to do what we would like to do. With some adaptation we can see if it yields good result.

**Comment by killerducky** [comment]
Some quotes from the paper about symmetry breaking:
> Figure 2: The Net2WiderNet transformation.... For a practical application, we would simultaneously replicate many randomly chosen units, and we would add a small amount of noise to break symmetry after the replication.

> When using Net2DeeperNet, there is no need to add noise. We can begin training with a student network that represents exactly the same function as the teacher network. Nevertheless, a small amount of noise can be added to break the symmetry more rapidly.

I skimmed the code in https://github.com/paengs/Net2Net and I couldn't tell right away if they are doing this. The rand calls around https://github.com/paengs/Net2Net/blob/master/net2net.py#L132 seem to be picking random nodes to replicate, but I don't see any random noise added.

**Comment by Mardak** [comment]
Reading a bit more discussions and usage of Net2Net, it sounds like it's pretty simple idea that is not difficult to implement or replicate.

Daniel Slater has a [numpy implementation](https://github.com/DanielSlater/Net2Net) and wrote:
>  Net2DeeperNet is quite simple, it creates an identity layer, then adds a small amount of noise. This means that the network activation is only unchanged if the layer is a linear layer, because otherwise the activation functions non-linearity will alter the output. So bare in mind if you have an activation function on your new layer(and you almost certainly will) then the network output will be changed and will have worse performance until it has gone through some amount of training.

http://www.danielslater.net/2016/05/using-net2net-to-speed-up-network.html
https://github.com/DanielSlater/Net2Net/blob/master/net2net/net_2_deeper_net.py

And here's an experience recapping usage of Net2Net for new research:
> We use Net2Net to reduce the training time in our modest computing facility and benefit from Siamese training without any architectural deficit. We apply Net2Net operators once in everytime training stalls through Siamese traning. In the end of the Siamese training we applied Net2Net wider operation once more to double the size and increase model capability to learn more representation.

> Wider operation adds more units to a layer by copying weights from the old units and normalizes the next layer weights by the cloning factor of each unit, in order to keep the propagated activation the same.  Deeper operation adds an identity layer between successive layers so that again the propagated activation stands the same.

> One subtle difference in our use of Net2Net is to apply zeroing noise to cloned weights in wider operation. It basically breaks the symmetry and forces each unit to learn similar but different representations.

https://badripatro.wordpress.com/2017/02/03/triplet-audio-and-video-and-tracking/#attachment_1440

**Comment by alreadydone** [comment]
If other prospective methods fail or are difficult to use, how about trying the following: let the 6x128 network (randomly initialized or trained from 5x64 self-play games) play with itself (or with the best 5x64 network, using pure networks) to generate positions, and use the output of the 5x64 network on these positions to train the 6x128 network? Just keep generating and training until a promotion occurs, then use the new best network to generate positions; repeat and stop when we get to a network roughly of same strength as the best 5x64.
The idea is to use the output of the stronger 5x64 network to correct misbehavior of the 6x128 network instead of using the winners of self-play games, saving computing power by not doing MCTS. If this method (or any other migration method) works well, we could increase the number of blocks one at a time instead of jump up to 10 blocks directly.

**Comment by gcp** [comment]
If someone manages to get a larger network with any of these methods that looks promising, let me know and I will queue it for matches.

**Comment by marcocalignano** [comment]
@killerducky I think we can really try https://github.com/paengs/Net2Net, in _wider_conv you give some bias as parameter.
But since I am not good in python (I bought a book and have to learn :smile:  ) I am stuck converting our weigth files to numpy matrixes and back.

**Comment by tapsika** [comment]
I don't know how relevant this is for go (probably not), but a while ago I tried something like this for a mnist test. With simply adding an extra conv layer to a half-trained net the avg loss did increase a little, even with plain relus and the weights initialized to "copy" (a single 1 and the rest to near-zero random for all filters). It OC increased more with normally initialized weights, but not as bad as I expected. Just from memory, something like accuracy 98% -> 97% vs 96% (that set is so easy it adjusts fast, and too deep nets are a bit worse there even when trained from scratch).

Also note the "results" section at https://github.com/paengs/Net2Net - it seems even with their method the difference to random inits is not that big after some adjustments. I haven't tried adding extra filters - this may be more promising (besides splitting old nodes to two halves, one may also try simply initing new weights/connections to near-zero random).

But I think an extra % in mnist can make a big difference in go, and what is a quick adjustment there can mean half of the total training here. So these techniques can probably only offer a small bonus, and not a substitute for significant retraining.

**Comment by odeint** [comment]
By now I also think retraining for scratch is the best. We'd do it anyway as a comparison, to check if the network resizing has a negative long-term effect. So why not start with it then, and give ourselves more time to think about (and implement...) the shortcuts properly.

**Comment by killerducky** [comment]
I looked a little more into https://github.com/paengs/Net2Net and the paper. It looks like the implementation assumes you only have convolutional or FC layers. It doesn't deal with batchnorm layers or skip connections. So it would take some more work to use this method.

It would probably take me a couple weekends to get this working. It seems interesting but I think I'll work on other things instead. Maybe someone else can take a crack at this.

**Comment by Mardak** [comment]
@paengs could you provide some assistance with applying Net2Net to this "open source AlphaGo Zero" project?

**Comment by paengs** [comment]
@Mardak Sure! How can I help AlphaGo Zero project? First, let me try to follow-up this projects.

**Comment by Mardak** [comment]
@gcp @marcocalignano or anyone else help get @paengs the context needed to dig into using something like Net2Net to improve the network?

@paengs Leela Zero so far has been mainly using 5 residual layers with 64 convolutional filters (5x64), and you can see the progression of Elo rating here: http://zero.sjeng.org/ The rate of Elo improvements per self-play games has been slowing down, so people, who have been contributing their GPU/CPUs in this public distributed effort, have been asking about making progress towards AlphaGo Zero's 40x256. There's nearly 3 million self-play games from various "best" networks for use in training, so one path of increasing the size of the network would be to train using those existing games, but so far that hasn't been successful.

This issue was opened to look at better ways to move from the 5x64 to a larger network, which brought us to Net2Net and your implementation. Where ideally we can apply that multiple times as Elo rating improvements slow down instead of starting from scratch.

Here's weights information from the README: https://github.com/gcp/leela-zero#weights-format

**Comment by gcp** [comment]
If you can read TensorFlow, the description is here:
https://github.com/gcp/leela-zero/blob/7f9489ee1bc870ade3ee8c9db8a61d7977528304/training/tf/tfprocess.py#L309

In words, it's:

- Convolutional 3x3 layer, BatchNorm, ReLU
- 5 x Residual blocks of (Conv 3x3 layer, BN, ReU, Conv 3x3, BN + Input, ReLU)
- Split output:
- FC layer of 361x256 => FC layer of 256x1 => 1 x TanH
- FC layer of 361x362 => 362 x Softmax

**Comment by tapsika** [comment]
FWIW out of curiosity I repeated my old hack/experiment, but this time with adding extra filters to conv layers. This does seem to work to an extent.

I trained a net on mnist for one pass (2 conv + 2 fc - no batchnorm). Then created and initialized a new net (with twice as much filters), and multiplied all weights with a constant C. Then copied the weights from the old net into their respective counterparts (new connections remained at init * C). Then continued training, looking at both the immediate loss/accuracy changes and the final performance after the second pass.

The behaviour depends on C of course. If it is too low, the new net will behave identically to the smaller one (loss not increasing, but only old parts work, new parts cannot learn, or take ages to start learning - like a bad init). If it is too high, the whole thing gets closer to a random init. I originally expected values around 1/10 to be useful, but at that levels I still got identical behaviour (maybe it would start learning after dozens or hundreds of passes).

It only started to react and outperform the old one at 1/1 levels (after a minimal immediate accuracy decrease) - but this probably depends on the dataset and net structure (and amount of training after the switch). Anyway, after the second pass, it was still worse than if the bigger net would have been trained from the start, but better than the smaller one.
